engine/pike2/pike2.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// calculateHMAC calculates the HMAC of the public shared secret using a reference string and an index.
+// The reference string and index are concatenated to form the HMAC key.
+// Returns the HMAC result as a byte slice or an error if the HMAC calculation fails.
+func calculateHMAC(pubSharedSecret []byte, reference string, idx int) ([]byte, error) {
+	h := hmac.New(sha256.New, []byte(fmt.Sprintf("%s-%d", reference, idx)))
+	if _, err := h.Write(pubSharedSecret); err != nil {
+		return nil, fmt.Errorf("error writing HMAC message - %w", err)
+	}
+	return h.Sum(nil), nil
+}
+
+// calculateDedicatedPublicKey calculates the dedicated public key (dPK) using the HMAC result and the receiver's public key.
+// The HMAC result is used as a scalar to perform elliptic curve scalar multiplication and point addition.
+// Returns the resulting public key or an error if the calculation fails.
+func calculateDedicatedPublicKey(hmacResult []byte, receiverPubKey *bec.PublicKey) (*bec.PublicKey, error) {
+	if len(hmacResult) == 0 {
+		return nil, fmt.Errorf("HMAC result is empty")
+	}
+	if receiverPubKey == nil {
+		return nil, fmt.Errorf("receiver public key is nil")
+	}
+
+	// Convert HMAC result to a big integer
+	hn := new(big.Int).SetBytes(hmacResult)
+	curve := bec.S256()          // Use secp256k1 curve
+	hn.Mod(hn, curve.Params().N) // Ensure the scalar is within the curve order
+
+	// Perform scalar multiplication: hn * G
+	rx, ry := curve.ScalarBaseMult(hn.Bytes())
+
+	// Perform point addition: (receiverPubKey + (hn * G))
+	dedicatedPubKeyX, dedicatedPubKeyY := curve.Add(receiverPubKey.X, receiverPubKey.Y, rx, ry)
+
+	// Verify that the resulting point is on the curve
+	if !curve.IsOnCurve(dedicatedPubKeyX, dedicatedPubKeyY) {
+		return nil, fmt.Errorf("resulting public key is not on curve")
+	}
+
+	// Create the dedicated public key
+	dPK := &bec.PublicKey{
+		Curve: curve,
+		X:     dedicatedPubKeyX,
+		Y:     dedicatedPubKeyY,
+	}
+	return dPK, nil
+}
+
+// convertToCompressedPublicKey converts the dedicated public key to its compressed format
+// and then hashes it using SHA-256 followed by RIPEMD-160 (Hash160).
+// Returns the compressed and hashed public key as a byte slice.
+func convertToCompressedPublicKey(dPK *bec.PublicKey) []byte {
+	// Serialize the public key to compressed format
+	dPKBytes := dPK.SerialiseCompressed()
+
+	// Apply Hash160 (SHA-256 followed by RIPEMD-160) to the compressed public key
+	dPKBytesCompressed := crypto.Hash160(dPKBytes)
+
+	// Return the compressed and hashed public key
+	return dPKBytesCompressed
+}
+
+// GenerateLockingScriptsFromTemplates converts Pike outputs templates to scripts
+func GenerateLockingScriptsFromTemplates(outputsTemplate []PikeOutputsTemplate, senderPubKey, receiverPubKey *bec.PublicKey, reference string) ([]string, error) {
+	lockingScripts := make([]string, len(outputsTemplate))
+	parser := interpreter.DefaultOpcodeParser{}
+	pubSharedSecret := CalculatePubSharedSecret(senderPubKey)
```

Comment:
As above. This step makes no sense as the output is exactly the same as senderPubKey.